crates/bevy_ecs/src/storage/thin_array_ptr.rs

```diff
@@ -0,0 +1,297 @@
+use crate::query::DebugCheckedUnwrap;
+use std::alloc::{alloc, handle_alloc_error, realloc, Layout};
+use std::mem::{needs_drop, size_of};
+use std::num::NonZeroUsize;
+use std::ptr::{self, NonNull};
+
+/// Similar to [`Vec<T>`], but with the capacity and length cut out for performance reasons.
+///
+/// This type can be treated as a `ManuallyDrop<Box<[T]>>` without a built in length. To avoid
+/// memory leaks, [`drop`](Self::drop) must be called when no longer in use.
+pub struct ThinArrayPtr<T> {
+    data: NonNull<T>,
+    #[cfg(debug_assertions)]
+    capacity: usize,
+}
+
+impl<T> ThinArrayPtr<T> {
+    fn empty() -> Self {
+        #[cfg(debug_assertions)]
+        {
+            Self {
+                data: NonNull::dangling(),
+                capacity: 0,
+            }
+        }
+        #[cfg(not(debug_assertions))]
+        {
+            Self {
+                data: NonNull::dangling(),
+            }
+        }
+    }
+
+    #[inline(always)]
+    fn set_capacity(&mut self, _capacity: usize) {
+        #[cfg(debug_assertions)]
+        {
+            self.capacity = _capacity;
+        }
+    }
+
+    /// Create a new [`ThinArrayPtr`] with a given capacity. If the `capacity` is 0, this will no allocate any memory.
+    #[inline]
+    pub fn with_capacity(capacity: usize) -> Self {
+        let mut arr = Self::empty();
+        arr.set_capacity(capacity);
```

Comment:
Do we need to call this? `alloc` calls it and it defaults to `0` (the only instance we don't call `alloc`)